File: limitless/src/main/java/user11681/limitless/enchantment/EnchantmentUtil.java
package user11681.limitless.enchantment;

import java.util.List;
import net.minecraft.enchantment.Enchantment;
import net.minecraft.enchantment.EnchantmentLevelEntry;
import net.minecraft.item.EnchantedBookItem;
import net.minecraft.item.ItemStack;
import net.minecraft.item.Items;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.nbt.ListTag;
import net.minecraft.util.Identifier;
import net.minecraft.util.registry.Registry;
import user11681.limitless.asm.access.EnchantmentAccess;

public interface EnchantmentUtil {
    String INTERNAL_NAME = "user11681.limitless.enchanting";

    int SUCCESS = 0;
    int ADD = 1;
    int CONFLICT = 2;

    static void getHighestSuitableLevel(final int power, final Enchantment enchantment, final List<EnchantmentLevelEntry> entries) {
        boolean found = false;
        int lastCandidate = 0;

        for (int i = enchantment.getMinLevel(); i <= enchantment.getMaxLevel(); i++) {
            if (power >= enchantment.getMinPower(i) && power <= enchantment.getMaxPower(i)) {
                lastCandidate = i;
                found = true;

                if (((EnchantmentAccess) enchantment).limitless_getOriginalMaxLevel() == 1) {
                    break;
                }
            } else {
                break;
            }
        }

        if (found) {
            entries.add(new EnchantmentLevelEntry(enchantment, lastCandidate));
        }
    }

    static void mergeEnchantment(final ItemStack itemStack, final EnchantmentLevelEntry enchantment, final boolean mergeConflicts) {
        final boolean book = itemStack.getItem() == Items.ENCHANTED_BOOK;

        if (book || itemStack.hasEnchantments()) {
            if (tryMerge(itemStack, enchantment, mergeConflicts) == ADD) {
                if (book) {
                    EnchantedBookItem.addEnchantment(itemStack, enchantment);
                } else {
                    itemStack.addEnchantment(enchantment.enchantment, enchantment.level);
                }
            }
        } else {
            itemStack.addEnchantment(enchantment.enchantment, enchantment.level);
        }
    }

    static void mergeEnchantment(final ItemStack itemStack, final Enchantment enchantment, final int level, final boolean mergeConflicts) {
        final boolean book = itemStack.getItem() == Items.ENCHANTED_BOOK;

        if (book || itemStack.hasEnchantments()) {
            if (tryMerge(itemStack, enchantment, level, mergeConflicts) == ADD) {
                if (book) {
                    EnchantedBookItem.addEnchantment(itemStack, new EnchantmentLevelEntry(enchantment, level));
                } else {
                    itemStack.addEnchantment(enchantment, level);
                }
            }
        } else {
            itemStack.addEnchantment(enchantment, level);
        }
    }

    static int tryMerge(final ItemStack itemStack, final EnchantmentLevelEntry enchantment, final boolean mergeConflicts) {
        return tryMerge(itemStack, enchantment.enchantment, enchantment.level, mergeConflicts);
    }

    static int tryMerge(final ItemStack itemStack, final Enchantment enchantment, final int level, final boolean mergeConflicts) {
        final boolean book = itemStack.getItem() == Items.ENCHANTED_BOOK;
        final ListTag enchantments;

        if (book) {
            enchantments = EnchantedBookItem.getEnchantmentTag(itemStack);
        } else {
            enchantments = itemStack.getEnchantments();
        }

        int status = ADD;

        for (final CompoundTag enchantmentTag : (Iterable<CompoundTag>) (Object) enchantments) {
            if (new Identifier(enchantmentTag.getString("id")).equals(Registry.ENCHANTMENT.getId(enchantment))) {
                final int tagLevel = enchantmentTag.getInt("lvl");

                if (tagLevel == level) {
                    enchantmentTag.putInt("lvl", Math.min(tagLevel + 1, enchantment.getMaxLevel()));
                } else {
                    enchantmentTag.putInt("lvl", Math.max(tagLevel, level));
                }

                return SUCCESS;
            } else if (!mergeConflicts && status != CONFLICT && !enchantment.canCombine(Registry.ENCHANTMENT.get(new Identifier(enchantmentTag.getString("id"))))) {
                status = CONFLICT;
            }
        }

        return status;
    }
}

File: limitless/src/main/kotlin/user11681/limitless/enchanting/EnchantmentUtil.kt
@file:Suppress("UNCHECKED_CAST")

package user11681.limitless.enchanting

import net.minecraft.enchantment.Enchantment
import net.minecraft.enchantment.EnchantmentLevelEntry
import net.minecraft.item.EnchantedBookItem
import net.minecraft.item.ItemStack
import net.minecraft.item.Items
import net.minecraft.nbt.CompoundTag
import net.minecraft.nbt.ListTag
import net.minecraft.util.Identifier
import net.minecraft.util.registry.Registry
import user11681.limitless.asm.access.EnchantmentAccess
import kotlin.math.max
import kotlin.math.min

object EnchantmentUtil {
    const val INTERNAL_NAME = "user11681.limitless.enchanting"
    const val DESCRIPTOR = "Luser11681.limitless.enchanting;"

    const val SUCCESS = 0
    const val ADD = 1
    const val CONFLICT = 2

    @JvmStatic
    fun getHighestSuitableLevel(power: Int, enchantment: Enchantment, entries: MutableList<EnchantmentLevelEntry>) {
        var found = false
        var lastCandidate = 0

        for (i: Int in enchantment.minLevel..enchantment.maxLevel) {
            if (power >= enchantment.getMinPower(i) && power <= enchantment.getMaxPower(i)) {
                lastCandidate = i
                found = true

                if ((enchantment as EnchantmentAccess).limitless_getOriginalMaxLevel() == 1) {
                    break
                }
            } else {
                break
            }
        }

        if (found) {
            entries.add(EnchantmentLevelEntry(enchantment, lastCandidate))
        }
    }

    @JvmStatic
    fun mergeEnchantment(itemStack: ItemStack, enchantment: EnchantmentLevelEntry, mergeConflicts: Boolean) {
        val book: Boolean = itemStack.item == Items.ENCHANTED_BOOK

        if (book || itemStack.hasEnchantments()) {
            if (tryMerge(itemStack, enchantment, mergeConflicts) == ADD) {
                if (book) {
                    EnchantedBookItem.addEnchantment(itemStack, enchantment)
                } else {
                    itemStack.addEnchantment(enchantment.enchantment, enchantment.level)
                }
            }
        } else {
            itemStack.addEnchantment(enchantment.enchantment, enchantment.level)
        }
    }

    @JvmStatic
    fun mergeEnchantment(itemStack: ItemStack, enchantment: Enchantment, level: Int, mergeConflicts: Boolean) {
        val book: Boolean = itemStack.item == Items.ENCHANTED_BOOK

        if (book || itemStack.hasEnchantments()) {
            if (tryMerge(itemStack, enchantment, level, mergeConflicts) == ADD) {
                if (book) {
                    EnchantedBookItem.addEnchantment(itemStack, EnchantmentLevelEntry(enchantment, level))
                } else {
                    itemStack.addEnchantment(enchantment, level)
                }
            }
        } else {
            itemStack.addEnchantment(enchantment, level)
        }
    }

    @JvmStatic
    fun tryMerge(itemStack: ItemStack, enchantment: EnchantmentLevelEntry, mergeConflicts: Boolean): Int {
        return tryMerge(itemStack, enchantment.enchantment, enchantment.level, mergeConflicts)
    }

    @SuppressWarnings("unchecked")
    @JvmStatic
    fun tryMerge(itemStack: ItemStack, enchantment: Enchantment, level: Int, mergeConflicts: Boolean): Int {
        var status: Int = ADD
        val book: Boolean = itemStack.item == Items.ENCHANTED_BOOK

        val enchantments: ListTag = if (book) {
            EnchantedBookItem.getEnchantmentTag(itemStack)
        } else {
            itemStack.enchantments
        }

        for (enchantmentTag: CompoundTag in enchantments as Iterable<CompoundTag>) {
            if (Identifier(enchantmentTag.getString("id")) == Registry.ENCHANTMENT.getId(enchantment)) {
                val tagLevel: Int = enchantmentTag.getInt("lvl")

                if (tagLevel == level) {
                    enchantmentTag.putInt("lvl", min(tagLevel + 1, enchantment.maxLevel))
                } else {
                    enchantmentTag.putInt("lvl", max(tagLevel, level))
                }

                return SUCCESS
            } else if (!mergeConflicts && status != CONFLICT && !enchantment.canCombine(Registry.ENCHANTMENT.get(Identifier(enchantmentTag.getString("id"))))) {
                status = CONFLICT
            }
        }

        return status
    }
}
